packages/ember-data/lib/system/relationships/has-many.js

```diff
@@ -144,9 +144,14 @@ export default function hasMany(type, options) {
       return relationship.getRecords();
     },
     set: function(key, records) {
+      Ember.assert("You must pass an array of records to set a hasMany relationship (elements must be instances of DS.Model)", (function() {
+        if (!isArrayLike(records)) { return false; }
+
+        return Ember.A(records).every((record) => Model.detectInstance(record) );
+      })());
```

Comment:
I am not sure how helpful this assertion message is (considering the use case in #3833). Should I add an explicit check if a `DS.PromiseObject` is present in the passed array, and if so, warn that the passed array contains the result of a `store.findRecord`/`store.queryRecord` which is not supported and how it should be resolved?